src/wp-admin/includes/file.php

```diff
@@ -684,7 +684,21 @@ function wp_tempnam( $filename = '', $dir = '' ) {
 	// Suffix some random data to avoid filename conflicts.
 	$temp_filename .= '-' . wp_generate_password( 6, false );
 	$temp_filename .= '.tmp';
-	$temp_filename  = $dir . wp_unique_filename( $dir, $temp_filename );
+	$temp_filename  = wp_unique_filename( $dir, $temp_filename );
+
+	/*
+	 * Filesystems typically have a limit of 255 characters for a filename.
+	 *
+	 * If the generated unique filename exceeds this, truncate the initial
+	 * filename and try again.
+	 */
+	$characters_over_255 = strlen( $temp_filename ) - 255;
+	if ( $characters_over_255 > 0 ) {
+		$filename = substr( $filename, 0, -$characters_over_255 );
+		return wp_tempnam( $filename, $dir );
+	}
```

Comment:
This is performed on the generated unique name rather than the original `$filename` because the number of characters added for uniqueness may change. Therefore, simply truncating the original `$filename` at the beginning of the function would be insufficient for futureproofing, or for files where  `(string) (255) "...filename9"` exists, and the unique name is `(string) (256) "...filename10"`.

Reply:
It may be preferable to patch `wp_unique_filename()`. However, `wp_unique_filename()` is used within the Media component.

This may result in:
- An existing `(string) (255) ...image-file.jpg` leading to a unique name of `(string) (257) ...image-file-1.jpg`.
- This would be truncated to `(string) (253) ...image-fil.jpg`.
- If this file exists, then `(string) (255) ...image-fil-1.jpg` would be generated.
- This would look confusing as it looks like a later version of `image-fil.jpg`, not `image-file.jpg`.

In addition, patching `wp_unique_filename()` instead would mean facilitating `file-with.ext` as well as `.hidden_file`, and truncating appropriately.

For the reported problem, this seems to be a low-impact solution for `wp_tempnam()`. Further  exploration could be carried out on `wp_unique_filename()` in a separate ticket, which would have input from Media component maintainers to explore potential problems this may cause.